### PR TITLE
Dimensions

### PR DESCRIPTION
**Pull request by shylux**
Original issue: https://github.com/4teamwork/pzbs.shop/issues/27

This PR adds a new option to specify the dimensions of a shop item, eg. the length of a piece of fabric.
There is a set of selectable dimensions as requested by our client, but it can be extended to any combination of units.
The dimensions are applied to the price of the shop item. For instance if you have fabric and the user selects 400mm the price will be multiplied by 400. So you have to specify the price by mm.

The dimensions have to be specified by the customer upon adding an item to the cart. They can be edited in the cart edit view. They are displayed in the order overview and specified in all emails. They also show up in the order manager and the csv export.

Also the "normal" view (contrary to the compact view) of a shop item has been removed. This has two reasons. First, the normal view was broken - you couldn't add an item to the cart. And second it would have been nigh impossible to add the dimensions without creating a ux atrocity. Keep in mind that the views are designed to work without javascript.
So the normal view has been replaced with the compact view.
screen shot 2017-03-27 at 11 11 07
screen shot 2017-03-27 at 11 12 06
screen shot 2017-03-27 at 11 13 09
screen shot 2017-03-27 at 11 13 37

### CODE REVIEW

File: ftw/shop/tests/test_shopitem_views.py [hunk 1]
```diff
@@ -48,15 +48,16 @@ def test_one_variation(self):
...
+        self.assertEquals(pq_doc("#var-Paperback td")[2].text, 'b22')
 
     def test_two_variations(self):
         item_datas = self.tshirt_view.getItemDatas()
         item = item_datas[0]
         item_listing = self.tshirt_view.two_variations(item)
         pq_doc = pq(item_listing)
-        self.assertEquals(len(pq_doc("input[name=skuCode][value='11']")), 1)
+        self.assertEquals(len(pq_doc('select[name=var1choice] option')), 3)
+        self.assertEquals(len(pq_doc('select[name=var2choice] option')), 3)
```
Comment (elioschmutz) [inline ftw/shop/tests/test_shopitem_views.py]: Why do you change that?

File: ftw/shop/tests/test_shopitem_views.py [hunk 2]
```diff
@@ -112,13 +113,9 @@ def test_variations_order(self):
...
-        all_sku_codes = pq_doc('input[name=skuCode]')
-        red_s = [e for e in all_sku_codes if e.value=='11'][0]
-        green_s = [e for e in all_sku_codes if e.value=='44'][0]
-        blue_s = [e for e in all_sku_codes if e.value=='77'][0]
-
-        self.assertTrue(all_sku_codes.index(red_s) < all_sku_codes.index(green_s) \
-            and all_sku_codes.index(green_s) < all_sku_codes.index(blue_s))
+        self.assertEquals(
+            ['Red', 'Green', 'Blue'],
+            [op.text for op in pq_doc('select[name=var1choice] option')])
```
Comment (elioschmutz) [inline ftw/shop/tests/test_shopitem_views.py]: Why do you change that?

File: ftw/shop/tests/test_shopitem_views.py [hunk 3]
```diff
@@ -129,14 +126,9 @@ def test_variations_order(self):
...
-        red_s = [e for e in all_sku_codes if e.value=='11'][0]
-        green_s = [e for e in all_sku_codes if e.value=='44'][0]
-        blue_s = [e for e in all_sku_codes if e.value=='77'][0]
-
-        self.assertTrue(all_sku_codes.index(red_s) > all_sku_codes.index(green_s) \
-            and all_sku_codes.index(green_s) > all_sku_codes.index(blue_s))
-
+        self.assertEquals(
+            ['Blue', 'Green', 'Red'],
+            [op.text for op in pq_doc('select[name=var1choice] option')])
```
Comment (elioschmutz) [inline ftw/shop/tests/test_shopitem_views.py]: Why do you change that?

File: ftw/shop/browser/templates/order_view.pt [hunk 4]
```diff
@@ -99,6 +99,7 @@
         <thead>
           <tr>
             <th i18n:translate="cart_header_product">Product</th>
+            <th i18n:translate="cart_header_dimensions">Dimensions</th>
```
Comment (elioschmutz) [inline ftw/shop/browser/templates/order_view.pt]: Do not include Dimensions if there are no shopitems with dimensions.

File: ftw/shop/browser/templates/order_view.pt [hunk 5]
```diff
@@ -108,6 +109,7 @@
           <metal:block tal:repeat="item python:order.cartitems">
             <tr>
               <td tal:content="item/title"></td>
+              <td><span metal:use-macro="context/@@shopitem_macros/dimensions_display"></span></td>
```
Comment (elioschmutz) [inline ftw/shop/browser/templates/order_view.pt]: Do not include Dimensions if there are no shopitems with dimensions.

File: ftw/shop/tests/test_cart.py [hunk 6]
```diff
@@ -190,7 +191,7 @@ def test_checkout(self):
         cart.checkout()
 
         order = omanager.getOrders()[0]
-        self.assertEquals(order.total, Decimal('7.15'))
+        self.assertEquals(order.total, Decimal('14.30'))
```
Comment (elioschmutz) [inline ftw/shop/tests/test_cart.py]: You just extend existing tests, but you added a new feature. The current testcases are very large, but perhaps you can add own testscases where you just test your feature here.
(this comment is for all the tests added in this commit)

File: ftw/shop/tests/test_category_view.py [hunk 7]
```diff
@@ -36,16 +36,17 @@ def test_one_variation(self):
         self.assertTrue('Hardcover' in item_listing)
 
         pq_doc = pq(item_listing)
-        self.assertEquals(len(pq_doc("input[name=skuCode][value=b11]")), 1)
-        self.assertEquals(len(pq_doc("input[name=skuCode][value=b22]")), 1)
+        self.assertEquals(pq_doc("#var-Hardcover td")[2].text, 'b11')
+        self.assertEquals(pq_doc("#var-Paperback td")[2].text, 'b22')
```
Comment (shylux) [inline ftw/shop/tests/test_category_view.py]: The default view changed from normal to compact so i had to update those tests.

**Comment by shylux** [comment]
* Added description for dimensions field.
 * Only display the dimensions column if they are used. (Tests for checkout etc cover that)
 * Added explicit test for dimensions selection & edit.
 * **Move comment field to review step.** This is another requirement which wasn't originally included in this PR.

@elioschmutz

**Comment by elioschmutz** [comment]
The following functions are still untestet:

- calc_item_total
- calc_vat
- has_specified_dimensions
- has_order_dimensions
- cart_has_dimensions
- validate_dimensions

You added some explicit dimension-tests implemented as a browser test. Thats good for an overall test. But to test the functions above, you should add functional-tests with all possible variations and all if/else cases. 

Some test examples:

```python
shopping_cart = IShoppingCart(obj, request)
self.assertEqual(15, shopping_cart.calc_item_total(15, 1, [1]), "Price should be respected in the calculation")
self.assertEqual(30, shopping_cart.calc_item_total(15, 2, [1]), "Quantity should be respected in the calculation")
self.assertEqual(30, shopping_cart.calc_item_total(15, 1, [2]), "Dimension should be respected in the calculation")
self.assertEqual(60, shopping_cart.calc_item_total(15, 1, [1, 2, 2]), "Multiple Dimensions should be respected in the calculation")
```

**Comment by shylux** [comment]
@elioschmutz 
Changed description to clarify price calculation.
screen shot 2017-04-04 at 10 48 34

It should be clear enough on the item view:
screen shot 2017-04-04 at 11 39 14

Added tests for dimension validation and new `calc_item_total` method. I didn't test the other methods because i think they are sufficiently covered by the integration tests.

**Pull request merged**